R/simulateGLV.R

```diff
@@ -1,98 +1,125 @@
 #'
 #' @examples
-#' row_data <- data.frame(Kingdom = "Animalia",
-#'                 Phylum = rep(c("Chordata", "Echinodermata"), c(500, 500)),
-#'                 Class = rep(c("Mammalia", "Asteroidea"), each = 500),
-#'                 ASV = paste0("X", seq_len(1000)),
-#'                 row.names = rownames(paste0("species", seq_len(1000))),
-#'                 stringsAsFactors = FALSE)
-#'
-#' row_data <- t(row_data)
-#'
-#' col_data <- DataFrame(sampleID = seq_len(1000),
-#'                     time = as.Date(1000, origin = "2000-01-01"),
-#'                     row.names = colnames(paste0("sample", seq_len(1000))))
-#'
 #' A <- miaSim::powerlawA(4, alpha = 1.01)
 #'
-#' SEobject <- simulateGLV(n.species = 4, A, t.start = 0, t.store = 1000)
-#' rowData(SEobject) <- row_data
-#' colData(SEobject) <- col_data
+#' ExampleGLV <- simulateGLV(n.species = 4, A, t.end = 1000)
 #'
-#' @docType methods
-#' @aliases simulateGLV-numeric
-#' @aliases simulateGLV,numeric-method
-#'
-#' @importFrom utils str
 #' @importFrom deSolve ode
 #' @importFrom stats runif
-#' @importFrom SummarizedExperiment SummarizedExperiment
-#' @importFrom methods setGeneric
 #'
 #' @export
-setGeneric("simulateGLV", signature = "n.species",
-    function(n.species, A, x = runif(n.species),
-            b = runif(n.species), t.start = 0, t.store, norm = FALSE, ...)
-        standardGeneric("simulateGLV"))
+simulateGLV <- function(n.species, A,
+        x = runif(n.species),
+        b = runif(n.species),
+        sigma.drift = 0.01,
+        sigma.epoch = 0.3,
+        sigma.external = 0.3,
+        p.epoch = 0.01,
+        t.external_events = c(12, 36, 48),
+        t.external_durations = c(3, 10, 99),
+        stochastic = FALSE,
+        norm = FALSE,
+        t.end = 1000, ...){
+
+
+        # input check
+        if(!isPositiveInteger(n.species)){
+            stop("n.species must be integer.")}
+        if(!all(vapply(list(A,x,b), is.double, logical(1)),
+                vapply(list(x,b), length, integer(1)) == n.species)){
+            stop("A,x,b must be matrix and the length must be equal to length
+                of n.species.")}
+
+        t.dyn <- simulationTimes(t.end = t.end, ...)
+        tEvent <- eventTimes(
+            t.events = t.external_events,
+            t.duration = t.external_durations, t.end = t.end, ...)
+        parameters <- list(b=b, A = A, n.species = n.species,
+            sigma.drift = sigma.drift, stochastic = stochastic,
+            sigma.epoch = sigma.epoch, p.epoch = p.epoch,
+            sigma.external = sigma.external, tEvent = tEvent)
+        out <- ode(
+            y = x,
+            times = t.dyn$t.sys,
+            func = dxdt,
+            parms = parameters,
+            events = list(func = perturb, time = t.dyn$t.sys))
+        counts <- t(out[,2:ncol(out)])
+        counts <- counts[,t.dyn$t.index]
+
+        if(norm){
+            counts <- t(t(counts)/colSums(counts))
```

Comment:
`colSums2` (and `colMeans2`) are optimized and faster, I propose to change here and elsewhere in the functions